src/cargo/sources/registry/index.rs

```diff
@@ -924,11 +954,20 @@ impl IndexSummary {
         }
         let mut summary = Summary::new(pkgid, deps, &features, links, rust_version)?;
         summary.set_checksum(cksum);
-        Ok(IndexSummary {
-            summary,
-            yanked: yanked.unwrap_or(false),
-            v,
-        })
+
+        let v_max = if bindeps {
+            INDEX_V_MAX + 1
+        } else {
+            INDEX_V_MAX
+        };
+
+        if v_max < v {
+            return Ok(IndexSummary::Unsupported(summary, v));
+        }
+        if yanked.unwrap_or(false) {
+            return Ok(IndexSummary::Yanked(summary));
+        }
+        Ok(IndexSummary::Candidate(summary))
```

Comment:
nit: since these are mutually exclusive branches, I would use an if / else-if setup instead to better communicate that to readers